src/core/thread/radio_selector.cpp

```diff
@@ -0,0 +1,420 @@
+            if (preference >= kHighPreference)
+            {
+                selectedRadio = radio;
+                break;
+            }
+
+            if (!found || (selectedPreference < preference))
+            {
+                found              = true;
+                selectedRadio      = radio;
+                selectedPreference = preference;
+            }
+        }
+    }
+
+    return selectedRadio;
+}
+
+Mac::TxFrame &RadioSelector::SelectRadio(Message &aMessage, const Mac::Address &aMacDest, Mac::TxFrames &aTxFrames)
+{
+    Neighbor *      neighbor;
+    Mac::RadioType  selectedRadio;
+    Mac::RadioTypes selections;
+
+    if (aMacDest.IsBroadcast() || aMacDest.IsNone())
+    {
+        aMessage.ClearRadioType();
+        ExitNow(selections.AddAll());
+    }
+
+    // If the radio type is already set when message was created we
+    // use the selected radio type. (e.g., MLE Discovery Response
+    // selects the radio link from which MLE Discovery Request is
+    // received.
+
+    if (aMessage.IsRadioTypeSet())
+    {
+        ExitNow(selections.Add(aMessage.GetRadioType()));
+    }
+
+    neighbor = Get<NeighborTable>().FindNeighbor(aMacDest, Neighbor::kInStateAnyExceptInvalid);
+
+    if ((neighbor == nullptr) || neighbor->GetSupportedRadioTypes().IsEmpty())
+    {
+        // If we do not have a corresponding neighbor or do not yet
+        // know the supported radio types, we try sending on all radio
+        // links in parallel. As an example, such a situation can
+        // happen when recovering a non-sleepy child (sending MLE
+        // Child Update Request to it) after device itself was reset.
+
+        aMessage.ClearRadioType();
+        ExitNow(selections.AddAll());
+    }
+
+    selectedRadio = Select(neighbor->GetSupportedRadioTypes(), *neighbor);
+    selections.Add(selectedRadio);
+
+    Log(OT_LOG_LEVEL_DEBG, "SelectRadio", selectedRadio, *neighbor);
+
+    aMessage.SetRadioType(selectedRadio);
+
+    // We (probabilistically) decide whether to probe on another radio
+    // link for the current frame tx. When probing we allow the same
+    // frame to be sent in parallel over multiple radio links but only
+    // care about the tx outcome (ack status) on the main selected
+    // radio link. This is done by setting the "required radio types"
+    // (`SetRequiredRadioTypes()`) to match the main selection on
+    // `aTxFrames`. We allow probe on TREL radio link if it is not
+    // currently usable (thus not selected) but is/was supported by
+    // the neighbor (i.e., we did rx/tx on TREL link from/to this
+    // neighbor in the past). The probe process helps detect whether
+    // the TREL link is usable again allowing us to switch over
+    // faster.
+
+#if OPENTHREAD_CONFIG_RADIO_LINK_TREL_ENABLE
+    if (!selections.Contains(Mac::kRadioTypeTrel) && neighbor->GetSupportedRadioTypes().Contains(Mac::kRadioTypeTrel) &&
+        (Random::NonCrypto::GetUint8InRange(0, 100) < kTrelProbeProbability))
```

Comment:
We would like to enable this in future, but I intentionally left it out for now from this PR (to reduce its complexity)
- The probing on 15.4 is much less useful, since if the neighbor link supports both TREL/WiFi and 15.4, we would in general prefer TREL for frame tx.
- We have the receive mechanism to discover the radio linkes (e.g. if/when a 15.4 link becomes available). For a router neighbor MLE Link Adv would be send on all radio links and potentially received by neighbors on multiple radio links as well). 
- Probing is mainly intended for cases where we are actively sending a lot of data/frames and want to quickly discover if a better radio link (TREL/WIFi in this case) has become available to switch to it quickly.

There are some other points to consider for probing on 15.4 links:
- When probing on 15.4 we don't want to use the same max tx retries (e.g., we use 15 tx attempts for direct tx). 
- Otherwise if the 15.4 link is not available, we would spend/waste a lot of time retx-ing the frame 15 times over the 15.4 link and waiting for ack timeout.
- So for this, we need to use mechanism to set a different tx retry count for probe frame vs normal data per radio type (we have this behavior in sub-MAC layer but need to make sure the radio platform also supports this to cover the cases where radio platform layer is handling the frame retx logic).